---
title: "bayesOmic: Association analysis of genomic data using a Bayesian shared component model"
author: "Laura Balagué, Juan J Abellán, Carlos Abellán, Juan R González*" 
date: "`r Sys.Date()`"
bibliography: bayesOmic.bib
output: 
  BiocStyle::html_document:
    number_sections: true
    toc: yes
    fig_caption: yes
    toc_float: true
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  message = FALSE, warning = FALSE,
  fig.height=16, fig.width=8
)
```
# Introduction
This document provides an overview of the `bayesOmic` package that will be available at Bioconductor (http://bioconductor.org/). The package implements a Bayesian approach for omic association studies. We propose a shared-component model to tease out the feature information (e.g. SNPs, CNVs, genes, transcripts, CpGs) that is common to cases and controls from the one that is specific to cases. This allows to detect the SNPs (bayesSNPassoc function) or any continuos feature (`bayesCNVassoc` function) that show the strongest association with the disease. The model can nevertheless be applied to more than one disease. More detailed information about the model and assumptions are given in @abellan2010bayesian and @gonzalez2012bayesian in the case of analyzing SNPs or CNVs, respectively. We illustrate how to analyze SNP data by using a synthetic data set (WTCCC will also be included). The simulated data set
contains information about 72 SNPs in 5 genes. It includes information for
800 individuals divided in 4 populations: controls and 3 type of cases (M1,
M2, and M3). We simulated SNPs from gene 1 to be associated with M1
and SNPs from gene 3 to M3.  
  
# Getting started
The `bayesOmic` package uses JAGS, a program for analysis of Bayesian hierarchical models using Markov Chain Monte Carlo (MCMC) simulation @plummer2003jags, to estimate model parameters. The current implementation of `bayesOmic` is based on JAGS version 4.3.0. JAGS has an R interface rjags that is used by the package (rjags version 4-9).

The development version of  `bayesOmic` can be installed from our GitHub repository

```{r install_bayesOmic, eval=FALSE}
library(devtools)
install_github("isglobal-brge/bayesOmic")
```

Then, the package is loaded as usual

```{r, load_bayesOmic}
library(bayesOmic)
```

# Analysis of SNP data

Let us illustrate how to apply our model to a GWAS study as described in @abellan2010bayesian. Let us start by describing how to get the data into R

## The data
Data can be imported from a text file or can be loaded using snpMatrix
package (to be supplied). We provide a simulated example that can be
loaded by typing:

```{r, load_sim_data}
data(sim.data)
```


This  is a simulated data having 72 SNPs and one column with case/control status

```{r show}
dim(sim.data)
sim.data[1:5, 1:5]
```

The package requires to have the case-control status and the SNPs in to
different objects:

```{r get model}
group <- sim.data$caco
SNPs <- sim.data[, -1]
```

It is worth to notice that GWAS data is normally in PLINK format. The Genomic Data Storage format is  .....   To be supplied (modify function)


## Model parameter estimates

The model can be fitted by using the function `bayesSNPassoc` as follows:

```{r run_SNP_model, eval=FALSE}
mod <- bayesSNPassoc(y=group, snps=SNPs)
```

This process takes about 5 minutes. To avoid waiting, we have saved the
object mod that can be loaded as:

```{r load_mod}
data(mod)
```

The function `bayesSNPassoc` prepares the data and then calls rjags to
estimate model parameters. We have set up the following default arguments
to be passed through rjags functions:

```{r}
args(bayesSNPassoc)
```

Notice that other arguments realated to MCMC estimation using JAGS can
be passed through this function. More details about them can be obtained
at http://calvin.iarc.fr/ martyn/software/jags/.  

The function also has two arguments to perform a simple quality control. The argument `call.rate` stands for the call rate of a given individual (e.g. proportion of non-missing SNPs, default is 0.9) and `min.freq` provides the minimum allele frequency allowed for a given SNP (default is 0.05). 

## Checking convergence
Before interpreting the simulations obtained from de a posteriori distri bution (e.g. model parameters), Markov chains convergence migth be verified. This can be done by using the function `checkConvergence`. This function has an argument called
`type` that defines the type of plot to be obtained. When `type="Markov
chain"` (default value) the function calls to `plot.mcmc()` function from package `coda` Ontherwise, Gelman-Rubin plots are displayed. The function `checkConvergence` has another argument, parameter, to indicate the model parameter to be summaryzed. The default is `alpha`. For example, Figure \ref{fig:checkConvergence} can be obtained by executing:

```{r checkConvergence, fig.cap="Chains convergence of `alpha` parameter for SNP model of 'sim.data'."}
checkConvergence(mod)
```

Other model parameters (Figure \ref{fig:checkConvergence2) are summaryzed by changing the argument called `parameter`.

```{r checkConvergence2, fig.cap="Chains convergence of `alpha` parameter for SNP model of 'sim.data'."}
checkConvergence(mod, parameter = "log-lambda")
```

Gelman-Rubin plot for `alpha` parameter can be obtained by typing

```{r checkConvergenceGR, fig.cap="Chains convergence of `alpha` parameter for SNP model of 'sim.data' using Gelman-Rubin method."}
checkConvergence(mod, type = "Gelman-Rubin")
```

## Results

Model parameters (intercept and shared component) can be obtained by
typing:

```{r showParam}
getParameters(mod)
```

On the other hand, specific and shared components can be visualize by:

```{r specificSNP}
plot(mod)
```

and

```{r sharedSNP}
plot(mod, type = "shared")
```

respectively.

Finally, a hierarchical clustering can be performed by using the predicted
probabilities by typing:

```{r}
library(RColorBrewer)
makeHeatmap(mod)
```

The figure shows a Heatmap were we can observe that groups M1 and M3 are
different from cases and group M1. 

##  Model Validation
DIC (Deviance Information Criteria) is one of the most used criteria to evaluate the goodness-of-fit for a given model. We use the function `dic.samples()` from rjags package to get such information. This deviance can be computed using two types of penalization depending on the argument type. The classic penalization proposed by @spiegelhalter2002bayesian can be obtained by setting `type="pD"`, while the penalization proposed by @plummer2008penalized can be obtained with `type="popt"`.

```{r dic, eval=FALSE}
dic.samples(mod$model, n.iter = 2000, type = "pD")
```

# Analysis of CNV data

We provide a simulated example that can be loaded by typing:

```{r}
data(armengol)
```

Multiple comparisons problem is address by computing confidence credible
intervals at more stringent level alpha.corrected

```{r}
nCNVs <- ncol(armengol) - 1
alpha.corrected <- (0.05/nCNVs)/2
```

Model parameter estimates are obtained using the function `bayesCNVassoc`
by executing:

```{r model_CNV, eval=FALSE}
mod.CNV <- bayesCNVassoc(armengol[, -1], armengol[, 1], method="JAGS",
                         alpha=c(alpha.corrected, alpha.corrected), 
                         n.iter.burn.in = 20000, n.iter = 200000)
```

This process is very time cosuming. We use INLA implementation instead

```{r model_CNV_inla}
mod.CNV <- bayesCNVassoc(armengol[, -1], armengol[, 1], method="inla",
                         alpha=c(alpha.corrected, alpha.corrected))
```

Specific components for each population can be obtained by typing:


```{r specificCNV}
plot(mod.CNV)
```

# 5 Acknowledgments
This work has been partly supported by ....  

# References